bundle/jsky.app.ot/src/main/scala/jsky/app/ot/ags/BagsManager.scala

```diff
@@ -65,6 +68,20 @@ object BagsState {
   type StateTransition = (BagsState, IO[Unit]) // Next BagsState, side-effects
   val ErrorTransition  = (ErrorState, ioUnit)  // Default transition for unexpected events
 
+  def hashObs(ctx: ObsContext): AgsHashVal = {
+    val when = ctx.getSchedulingBlockStart.asScalaOpt | Instant.now.toEpochMilli
+    AgsHash.hash(ctx, when)
+  }
+
+  def hashObs(o: ISPObservation): Option[AgsHashVal] =
+    ObsContext.create(o).asScalaOpt.map(hashObs)
+
+  def fetchObs(k: ObsKey): Option[ISPObservation] =
+    Option(SPDB.get().lookupProgram(k.prog.key)).flatMap { p =>
```

Comment:
Unrelated to this PR but is the call `SPDB.get().lookupProgram()` the canonical way to read the program from the `SPNodeKey`? would it read the local db only?